src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpointManagerBase.cs

```diff
@@ -23,12 +23,16 @@ internal abstract class ServiceEndpointManagerBase : IServiceEndpointManager
         // Filtered valuable endpoints from ServiceOptions, use dict for fast search
         public IReadOnlyDictionary<ServiceEndpoint, ServiceEndpoint> Endpoints { get; private set; }
 
+        private event EndpointEventHandler OnCreate;
         public event EndpointEventHandler OnAdd;
         public event EndpointEventHandler OnRemove;
         
-        protected ServiceEndpointManagerBase(IServiceEndpointOptions options, ILogger logger)
+        protected ServiceEndpointManagerBase(IServiceEndpointOptions options, IAccessKeyManager manager, ILogger logger)
             : this(ServiceEndpointUtility.Merge(options.ConnectionString, options.Endpoints), logger)
         {
+            OnAdd += (e) => manager.AddHubServiceEndpoint(e);
```

Comment:
As the endpoint changes always happen here, it seems that you just need to update the endpoint store in `AccessKeyManager` here. You can add endpoints to the manager before endpoints add to each hub and remove endpoints after them all removed from hub.
https://github.com/Azure/azure-signalr/blob/cff270e73cacaf55c3abd6aa73b0b7029b9c2f09/src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpointManagerBase.cs#L84

And you don't need to care about deduplication because `GetValuableEndpoints` already does this.
https://github.com/Azure/azure-signalr/blob/cff270e73cacaf55c3abd6aa73b0b7029b9c2f09/src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpointManagerBase.cs#L88